### PR TITLE
enh(UI): Add suggestions for dynamic criterias

### PR DESCRIPTION
**Pull request by Thebarda**
## Description

This adds suggestions for for dynamic criterias.
https://www.loom.com/share/93b9b050b99c46cc970df9d58710920a

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.04.x
- [ ] 20.10.x
- [ ] 21.04.x
- [x] 21.10.x (master)

<h2> How this pull request can be tested ? </h2>

- In the search bar
- Type 'hos'
- Press tab
- -> 'host_group:' is completed and some suggestions should appear
- Press arrow down / arrow up to select a suggestion
- -> The suggestion is applied in the search bar
- Add a ','
- Type something to search a specific host group
- Select another host group
- You can repeat repeat these steps for monitoring server and service group

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts [hunk 1]
```diff
@@ -214,18 +216,25 @@ const getCriteriaValueSuggestions = ({
...
-  const isNextCharacterEmpty =
-    isNil(nextCharacter) || isEmpty(nextCharacter.trim());
 
-  if (isNil(cursorPosition) || !isNextCharacterEmpty) {
-    return [];
-  }
+  return isNil(nextCharacter) || isEmpty(nextCharacter.trim());
+};
+
+interface GetCRiteriaAndExpression {
```
Comment (bdauria) [inline www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts:228]:
```suggestion
interface CriteriaExpression {
```

File: www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts [hunk 2]
```diff
@@ -214,18 +216,25 @@ const getCriteriaValueSuggestions = ({
...
+  criteriaName: string;
+  expressionBeforeCursor: string;
+  expressionCriteriaValues: Array<string>;
+}
 
+const getCriteriaAndExpression = ({
+  search,
+  cursorPosition,
+}: AutocompleteSuggestionProps): GetCRiteriaAndExpression => {
```
Comment (bdauria) [inline www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts:237]:
```suggestion
}: AutocompleteSuggestionProps): CriteriaExpression => {
```

File: www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts [hunk 3]
```diff
@@ -240,6 +249,67 @@ const getAutocompleteSuggestions = ({
...
   const expressionCriteriaValues = pipe(last, split(','))(expressionCriteria);
 
+  return {
+    criteriaName,
+    expressionBeforeCursor,
+    expressionCriteriaValues,
+  };
+};
+
+interface GetDynamicCriteriaParametersAndValueProps {
```
Comment (bdauria) [inline www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts:259]:
```suggestion
interface DynamicCriteriaParametersAndValues {
```

File: www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts [hunk 4]
```diff
@@ -240,6 +249,67 @@ const getAutocompleteSuggestions = ({
...
+  criteria: CriteriaDisplayProps;
+  values: Array<string>;
+}
+
+const getDynamicCriteriaParametersAndValue = ({
+  search,
+  cursorPosition,
+}: AutocompleteSuggestionProps): GetDynamicCriteriaParametersAndValueProps | null => {
```
Comment (bdauria) [inline www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts:267]:
```suggestion
}: AutocompleteSuggestionProps): DynamicCriteriaParametersAndValues | null => {
```

File: www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts [hunk 5]
```diff
@@ -268,4 +338,11 @@ const getAutocompleteSuggestions = ({
...
 };
 
-export { parse, build, getAutocompleteSuggestions, searchableFields };
+export {
+  parse,
+  build,
+  getAutocompleteSuggestions,
+  getDynamicCriteriaParametersAndValue,
+  searchableFields,
+  GetDynamicCriteriaParametersAndValueProps,
```
Comment (bdauria) [inline www/front_src/src/Resources/Filter/Criterias/searchQueryLanguage/index.ts:347]:
```suggestion
  DynamicCriteriaParametersAndValues,
```

**Review by bdauria — DISMISSED** [review]
LGTM and tested

**Review by Jordan91 — DISMISSED** [review]
works very well , nice work!

**Review by bdauria — APPROVED** [review]

**Review by jdelpierre — APPROVED** [review]

**Comment by jdelpierre** [comment]
Nice this :)

**Pull request merged**